app/src/components/pointcloud/PointCloudPointTooltip.tsx

```diff
@@ -0,0 +1,97 @@
+import React from "react";
+import { Html } from "@react-three/drei";
+import * as THREE from "three";
+import { css } from "@emotion/react";
+
+import { useDatasets, usePointCloudContext } from "@phoenix/contexts";
+import { getDatasetRoleFromEventId } from "@phoenix/utils/pointCloudUtils";
+
+import { EventItem } from "./EventItem";
+
+/**
+ * The size of the tooltip. E.g. 200px x 200px
+ */
+const TOOLTIP_SIZE = 200;
+/**
+ * The offset from the mouse position to the tooltip
+ */
+const TOOLTIP_OFFSET = 10;
+/**
+ * Re-used position vector
+ */
+const vec = new THREE.Vector3();
+export const PointCloudPointTooltip = () => {
+  const { primaryDataset, referenceDataset } = useDatasets();
+  const hoveredEventId = usePointCloudContext((state) => state.hoveredEventId);
+  const eventIdToDataMap = usePointCloudContext(
+    (state) => state.eventIdToDataMap
+  );
+  const pointData = usePointCloudContext((state) => state.pointData);
+  const pointGroupColors = usePointCloudContext(
+    (state) => state.pointGroupColors
+  );
+  const eventIdToGroup = usePointCloudContext((state) => state.eventIdToGroup);
+  if (hoveredEventId == null || pointData == null || pointData == null)
+    return null;
+
+  // The raw event from the first query
+  const baseEvent = eventIdToDataMap.get(hoveredEventId);
+  const eventDetails = pointData[hoveredEventId];
+  // Show nothing if everything is not loaded
+  if (baseEvent == null || eventDetails == null) return null;
+  const group = eventIdToGroup[hoveredEventId];
+  const color = pointGroupColors[eventIdToGroup[hoveredEventId]];
+  const datasetRole = getDatasetRoleFromEventId(hoveredEventId);
+  const datasetName =
+    datasetRole === "primary"
+      ? primaryDataset.name
+      : referenceDataset?.name ?? "reference";
+  return (
+    <Html
+      position={baseEvent.position}
+      pointerEvents="none"
+      zIndexRange={[0, 1]}
+      calculatePosition={(el, camera, size) => {
```

Comment:
@Parker-Stafford I figured out a way to not have to use debounce or timing I think. This makes the tooltip HTML be positioned away from the point and also detects the edges of the canvas.

Reply:
oh nice, will take a look

Reply:
I re-added debounce just to eliminate the frames dropping.